extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,389 @@
+   */
+  getAsyncCid_() {
+    this.dialogPromise_ = new Promise((resolve, reject) => {
+      this.dialogResolve_ = resolve;
+    });
+    return cidFor(this.win_).then(cid => {
+      // `amp-user-notification` is our cid scope, while we give it a resolved
+      // promise for the 2nd argument so that the 3rd argument (the
+      // persistentConsent) is the one used to resolve getting
+      // the external CID.
+      return cid.get('amp-user-notification',
+          Promise.resolve(), this.dialogPromise_);
+    });
+  }
+
+  /** @override */
+  shouldShow() {
+    return this.getAsyncCid_()
+        .then(this.postShowEndpoint_.bind(this))
+        .then(this.onPostShowEndpointSuccess_.bind(this));
+  }
+
+  /**
+   * Shows the current user notification. Adds the `.amp-active` class to
+   * the element.
+   */
+  show() {
+    this.element.style.display = 'block';
+    this.element.classList.add('amp-active');
+    return this.dialogPromise_;
+  }
+
+  /**
+   * Hides the current user notification and invokes the `dialogResolve_`
+   * method. Removes the `.amp-active` class from the element.
+   */
+  dismiss() {
+    this.element.style.display = 'none';
+    this.element.classList.remove('amp-active');
+    this.element.classList.add('amp-hidden');
+    this.dialogResolve_(false);
+    this.postDismissEnpoint_();
+  }
+}
+
+
+/**
+ * UserNotificationManager handles `amp-user-notification`
+ * queuing and registration, as well as exposing the components
+ * dismiss promise.
+ */
+export class UserNotificationManager {
+
+  constructor(window) {
+    this.win_ = window;
+
+    /** @private @const {!Object<!UserNotificationDeferDef>} */
+    this.deferRegistry_ = Object.create(null);
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win_);
+
+    /** @private @const {?function(*)} */
+    this.documentReadyResolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.documentReadyPromise_ = new Promise((resolve) => {
+      this.documentReadyResolve_ = resolve;
+    });
+
+    // TODO: Add `whenDocumentReady` function in document-state
+    // that returns a promise.
+    // See https://github.com/ampproject/amphtml/issues/1226
+    onDocumentReady(this.win_, () => {
+      this.documentReadyResolve_();
+    });
+
+    /** @private {!Promise} */
+    this.nextInQueue_ = this.viewer_.whenVisible().then(() => {
+      // we block on both viewer visible and when dom ready.
+      return this.documentReadyPromise_;
+    }).then(() => {
+      Object.keys(this.deferRegistry_).forEach((id) => {
+        assert(this.win_.document.getElementById(id) != null,
+            `Did not find amp-user-notification element ${id}`);
+      });
+    });
+  }
+
+  /**
+   * Retrieve a promise associated to an `amp-user-notification` component
+   * that is resolved when user agrees to the terms.
+   * @param {string} id
+   * @return {!Promise<string>}
+   */
+  get(id) {
+    assert(!!this.win_.document.getElementById(id),
```

Comment:
Sorry, meant to say that we actually need it, but with the caveat. If `get` is called before DOMReady (in other words before this service has started) - we can't really query DOM, so we can't perform this check, instead with put these requests into a queue. But the queue will be tested right after the service starts - so this is all good. However, if `get` is called after the service starts, it's both (a) we can query DOM and test it right away and (b) the check above on the queue will not be able to flag this for us.

I think we can find a better way to test this. I think the best place is to check is at the dequeue time. E.g. once we dequeue the next notification to process from the deferred queue - we can immediately test if it's in the DOM. That might be the easiest option.